### DISCUSSION TITLE
Creating Videos with Stellarium

### DISCUSSION THREAD

**Discussion by beltoforion** [Stellarium/stellarium#3638] [Show and tell]
I have recently updated a [python based script utility that allows to create videos with Stellarium](https://beltoforion.de/en/stellarium_video). The new version runs on Windows and Linux and is capable of using custom script templates.

[stellarium-to-video at GitHub](https://github.com/beltoforion/stellarium-to-video)

You can now use custom script templates and set the observer location to different planets. Creating a video can be done with a single command. 

# Example 1: Planetary alignment from 1961

```python
python3 stellarium-to-video.py -v180,30,70 -l Freiberg -c "Dance of Planets" -d1961-10-08T12:00:00+01:00 -ts P6M -s -o "dance.mp4" -dt P1D -tplanet_track.ssc -sz960x540
```
https://github.com/beltoforion/stellarium-to-video/assets/2202567/2faa0861-ecc2-484a-9a12-082c51b60166

# Example 2: Analemma 

```python
python3 stellarium-to-video.py -l Freiberg -t suntrack.ssc -c Analemma -d2024-06-20T12:00:00+01:00 -ts P1Y -s -o output.mp4 -dt P2D
```
https://github.com/beltoforion/stellarium-to-video/assets/2202567/c40f8c4f-f63a-4487-ad46-29602272e501

**Comment by alex-w** [comment]
Perfect! Would you write a chapter for Stellarium User Guide about using your script/tool with examples and information where it can be obtained, like the Scripting chapter? I think it may be very good addition to the Guide and I guess it should be put immediately after Scripting chapter.

**Comment by gzotti** [comment]
Very nice indeed! I usually just capture a running script with whatever screen grabber is available, but this has guaranteed speed. 

Maybe you can even remove the skyfield dependency here (but thanks for pointing me to it):  You can find the setting time of the sun simply by `core.getObjectInfo("sun")['set']`

Now the next challenge will be a purely python solution that even avoids the temporary script, controlling Stellarium via RemoteControl interface. And capturing scripts with sound output.

**Comment by beltoforion** [comment]
Yes I can write such a chapter in the documentation.  There is however one addition I''d like to make in the script. I'd like to be able to create videos in portrait orientation. (see [Issue 10 of stellarium-to-video](https://github.com/beltoforion/stellarium-to-video/issues/10)). 

Right now I could submit a config.ini via the command line to force stellarium to run in a window that has portrait mode orientation. However Stellarium will not let me create a window that exceeds the size of the screen. So any video in portrait mode would be limited to 1080 pixels in height on a Full-HD screen. I can manually drag the window to any size at runtime but I cannot do this via the config file.

There are a few ways around that:
- Allow Stellarium to open a window that exceeds the screen size when such a size is submitted in the config.ini.
- provide a way to control and change Stellariums window size from a script (i.e. adding a WindowMgr class)
- apart from this is could use 3rd party UI-automation tools like pywinauto for manipulating Stellariums window but they are not platform independent.

Maybe i can implement this myself. Would you be willing to merge a PR that either allows stellarium to create windows that exceed the screen dimension or add a class to allow stellarium scripts to alter the state of stellariums main window?

**Reply by alex-w** [reply]
Hmm… why you think that Stellarium cannot do screenshots over screen resolution?

**Reply by gzotti** [reply]
The screenshot size is limited by your GPU framebuffer limits. Just configure your size before and activate usage of the custom size, or add those settings to your custom config.ini. StelMainView (where these settings are) is currently not scriptable.

**Comment by beltoforion** [comment]
I tried setting up a window size of 1080x1920 in the "config.ini". However on startup the window will be clipped to the screen dimensions and I end up with appears to be a 1080x1080 window. This seems to be enforced by the Window Manager for events that come via X11/EWMH. Most likely Stellarium cannot do anything to prevent that because it is bound to the rules of the window manager. 

I found a way around it that at least works for GNOME but it is quite dirty and involves faking Atl+F8 keyboard events to move the window partially outside the screen which disables the window manager restrictions for resizing. It appears that you can manually resize a window to a dimension larger then the screen but you cannot do it programmatically via EWMH events at least not without reconfiguring the window manager.

With this hack i can fully automatically create videos in full hd portrait orientation or 4K videos on a Full-HD screen:

https://github.com/beltoforion/stellarium-to-video/assets/2202567/baee9422-7594-4fc3-a7c0-fe4562d6a518

**Reply by gzotti** [reply]
You can set custom dimensions of the screenshot (and activate them) in the configuration/Tools GUI. Don't forget to store settings then. Thanks for the hint about resizing beyond screen size on GNOME.

**Comment by beltoforion** [comment]
@gzotti Thank you for pointint this out. Setting the screenshot size directly would be so much easier than what I'm doing right now and it would work on all platforms. Can i set those parameters in the config.ini or from a stellarium script?

**Reply by gzotti** [reply]
I think it's not scriptable, but stored in config.ini.

**Reply by beltoforion** [reply]
Thanks for pointing this out. Having it in the config file would be sufficient as i can submit a config file when stellarium is started. However i dont't think the parameter is in there yet. At least i could not find it. There is this main section in my config:

[main]
invert_screenshots_colors              = false
screenshot_format                      = png
version                                = 23.4

Since the inverted_screenshot_colors option is there i think the resolution would also be there but it appears to be absent.

**Reply by alex-w** [reply]
Please check Stellarium User Guide, D.1.14

**Reply by gzotti** [reply]
please store your custom settings once and find e.g.:

```
[main]
invert_screenshots_colors                 = false
restore_defaults                          = false
screenshot_custom_height                  = 768
screenshot_custom_size                    = false
screenshot_custom_width                   = 1024
screenshot_datetime_filemask              = yyyyMMdd-hhmmssz
screenshot_datetime_filename              = false
screenshot_dir                            = 
screenshot_format                         = png

```

**Reply by beltoforion** [reply]
Thanks alot.
